lib/lcdbwf/R/helpers.R

```diff
@@ -493,59 +504,413 @@ nested.lapply <- function(x, subfunc, ...){
+
+    if(!file.exists(rds_file)){
+      stop(paste('RDS file does not exist:', rds_file))
+    }
+
+    # get res_list & dds_list (and any others) from RDS file
+    tmp <- readRDS(rds_file)
+
+    if(!any(c('res_list', 'dds_list') %in% names(tmp))){
+      stop('Object must contain "res_list" & "dds_list" elements!')
+    }
+
+    res_list <- tmp$res_list
+    dds_list <- tmp$dds_list
+
+    if('rld_list' %in% names(tmp)) rld_list <- tmp$rld_list
+    if('enrich_list' %in% names(tmp)) enrich_list <- tmp$enrich_list
+    if('degpatterns_list' %in% names(tmp)) degpatterns_list <- tmp$degpatterns_list
+    if('all_dds' %in% names(tmp)) all_dds <- tmp$all_dds
+    if('all_rld' %in% names(tmp)) all_rld <- tmp$all_rld
+  }
 
+  message('\n1. Processing res_list & dds_list')
   # Much of this function is just checking that the names all line up.
   res_dds_names <- unlist(lapply(res_list, function (x) x$dds))
   names(res_dds_names) <- NULL
   dds_dds_names <- names(dds_list)
   res_not_dds <- setdiff(res_dds_names, dds_dds_names)
   dds_not_res <- setdiff(dds_dds_names, res_dds_names)
   if (length(res_not_dds) > 0){
-    stop(paste("The following dds names are in res_list but are not found in dds_list:", res_not_dds, '\n'))
+    stop(paste("\t- The following dds names are in res_list but are not found in dds_list:",
+               paste(res_not_dds, collapse=', '), '\n'))
   }
+
+  # NOTE: drop unused dds_list & rld_list objects
   if (length(dds_not_res) > 0){
-    warning(paste("The following dds names are in dds_list but not in res_list. This OK, but may be unexpected:", dds_not_res, '\n'))
+    message("\t- The following dds names are in dds_list but not in res_list. These will be skipped:")
+    message(paste0('\t\t', paste(dds_not_res, collapse='\n\t\t')))
+    dds_list <- dds_list[ setdiff(dds_dds_names, dds_not_res) ]
+    if(!is.null(rld_list)){
+      rld_list <- rld_list[ setdiff(names(rld_list), dds_not_res) ]
+    }
   }
 
   # check if rld_list was specified, if not make it
   if(is.null(rld_list)){
+    message("\t- rld_list was not specified. Generating it")
     rld_list <- lapply(dds_list,
                   function(x) varianceStabilizingTransformation(x, blind=TRUE)
                 )
   }
 
-  obj <- list(
-    res_list=res_list,
-    dds_list=dds_list,
-    rld_list=rld_list
-  )
+  # sanitize res_list, dds_list & rld_list
+  obj <- sanitize_res_dds(res_list=res_list,
+                          dds_list=dds_list,
+                          rld_list=rld_list)
+
+  # if all_dds not specified, but dds_list has length 1,
+  # then use dds_list[[ 1 ]] as all_dds
+  if(is.null(all_dds) & length(obj$dds) == 1){
+    message('\t- all_dds was not specified, but dds_list has only 1 object. Using that instead')
+    all_dds <- obj$dds[[ 1 ]]
+
+    # if specifying all_dds, compute all_rld even if specified
+    all_rld <- varianceStabilizingTransformation(all_dds, blind=TRUE)
```

Comment:
Not obvious to me why it needs to be remade, I would think if you had a single dds you'd have a single rld? Either way, should at least give a message here because I would not expect a function to silently overwrite an arg I had provided.

Reply:
Agreed, there should at least be a message. Maybe `all_rld` shouldn't even be an argument, i.e. the function
will always compute this. I basically have this here to ensure a 1:1 mapping between `all_dds` and `all_rld`, which
might not be true if i allow independent objects to be specified. Basically, this protects from user error.